File: benches/suite/cpp/dijkstra.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, target_m, seed, num_q;
    if (!(cin >> n >> target_m >> seed >> num_q)) return 0;

    mt19937 rng(seed);
    vector<vector<pair<int, int>>> adj(n);

    for (int i = 0; i < n - 1; ++i) {
        int w = uniform_int_distribution<int>(1, 100)(rng);
        adj[i].push_back({i + 1, w});
        adj[i + 1].push_back({i, w});
    }

    int added = n - 1;
    while (added < target_m) {
        int u = uniform_int_distribution<int>(0, n - 2)(rng);
        int v = uniform_int_distribution<int>(u + 1, min(u + 50, n - 1))(rng);
        int w = uniform_int_distribution<int>(1, 100)(rng);
        adj[u].push_back({v, w});
        adj[v].push_back({u, w});
        added++;
    }

    const long long INF = numeric_limits<long long>::max();
    vector<long long> dist(n, INF);
    dist[0] = 0;
    priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
    pq.push({0, 0});

    while (!pq.empty()) {
        long long d = pq.top().first;
        int u = pq.top().second;
        pq.pop();

        if (d > dist[u]) continue;
        for (auto& edge : adj[u]) {
            int v = edge.first;
            int w = edge.second;
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }

    mt19937 rng2(seed + 1);
    for (int i = 0; i < num_q; ++i) {
        int t = uniform_int_distribution<int>(1, n - 1)(rng2);
        if (dist[t] == INF) {
            cout << -1 << "\n";
        } else {
            cout << dist[t] << "\n";
        }
    }

    return 0;
}

File: benches/suite/java/dijkstra.java
import java.util.*;
import java.io.*;

public class Solution {
    static class Edge {
        int to, weight;
        Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }

    static class Node implements Comparable<Node> {
        int u;
        long d;
        Node(int u, long d) {
            this.u = u;
            this.d = d;
        }
        public int compareTo(Node other) {
            return Long.compare(this.d, other.d);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line = br.readLine();
        if (line == null) return;
        StringTokenizer st = new StringTokenizer(line);
        int n = Integer.parseInt(st.nextToken());
        int targetM = Integer.parseInt(st.nextToken());
        int seed = Integer.parseInt(st.nextToken());
        int numQ = Integer.parseInt(st.nextToken());

        Random rng = new Random(seed);
        List<List<Edge>> adj = new ArrayList<>(n);
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());

        for (int i = 0; i < n - 1; i++) {
            int w = rng.nextInt(100) + 1;
            adj.get(i).add(new Edge(i + 1, w));
            adj.get(i + 1).add(new Edge(i, w));
        }

        int added = n - 1;
        while (added < targetM) {
            int u = rng.nextInt(n - 1);
            int v = rng.nextInt(Math.min(u + 50, n - 1) - (u + 1) + 1) + (u + 1);
            int w = rng.nextInt(100) + 1;
            adj.get(u).add(new Edge(v, w));
            adj.get(v).add(new Edge(u, w));
            added++;
        }

        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.add(new Node(0, 0));

        while (!pq.isEmpty()) {
            Node curr = pq.poll();
            if (curr.d > dist[curr.u]) continue;
            for (Edge e : adj.get(curr.u)) {
                if (dist[curr.u] + e.weight < dist[e.to]) {
                    dist[e.to] = dist[curr.u] + e.weight;
                    pq.add(new Node(e.to, dist[e.to]));
                }
            }
        }

        rng = new Random(seed + 1);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < numQ; i++) {
            int t = rng.nextInt(n - 1) + 1;
            if (dist[t] == Long.MAX_VALUE) {
                sb.append("-1\n");
            } else {
                sb.append(dist[t]).append("\n");
            }
        }
        System.out.print(sb.toString());
    }
}
